### PR TITLE
Add an option to select Spack's new concretizer

### PR DESCRIPTION
**Pull request by joshessman-llnl**
This approach uses the `json` config file to expose the clingo option, which triggers a modification of Spack's `config.yaml` in addition to an install of clingo through `pip`, packaged with https://github.com/joshessman-llnl/package-clingo-manylinux.

~I am in communication with the clingo dev(s) regarding an official PPC build that would allow us to switch over to the official package on PyPI.~ ✔️ This wheel-based approach was selected (instead of spack's bootstrap option) to avoid having to build clingo and its dependencies from source.

As an alternative interface, uberenv could read in the provided `config.yaml` (instead of the json option), see if it specifies the new concretizer, and install clingo accordingly.  Thoughts?

### CODE REVIEW

File: uberenv.py [hunk 1]
```diff
@@ -770,6 +776,17 @@ def patch(self):
                 print("[copying patched packages from {0}]".format(_src_glob))
                 sexe("cp -Rf {0} {1}".format(_src_glob, dest_spack_pkgs))
 
+        # Update spack's config.yaml if clingo was requested
+        if self.use_clingo:
+            import yaml
+            with open(pjoin(spack_etc_defaults_dir, "config.yaml"), 'r') as conf_file:
```
Comment (cyrush) [inline uberenv.py:782]: Good that this import is deferred b/c pyyaml is not a part of all standard python installs.

File: uberenv.py [hunk 2]
```diff
@@ -770,6 +776,17 @@ def patch(self):
...
+                conf_data = conf_file.read()
+            conf_yaml = yaml.load(conf_data)
+            conf_yaml['config']['concretizer'] = 'clingo'
+            # This will leave out all the comments, is this a problem?
+            with open(pjoin(spack_etc_defaults_dir, "config.yaml"), 'w') as conf_file:
```
Comment (cyrush) [inline uberenv.py:787]: this is a copy that is placed in the uberenv spack dir right? I think its fine if comments are stripped. 

Q: could you check if the file already has `clingo` specified, and if so skip the injection?

File: uberenv.py [hunk 3]
```diff
@@ -1035,6 +1052,39 @@ def use_spack_upstream(self):
...
+            py_interp = sys.executable
+            clingo_pkg = "clingo"
+            uninstall_cmd = "{0} -m pip uninstall -y {1}".format(py_interp, clingo_pkg)
+            # Uninstall it first in case the available version failed due to differing arch
+            # pip will still return 0 in the case of a "trivial" uninstall
+            res = sexe(uninstall_cmd, echo=True)
+            if res != 0:
+                print("[ERROR: clingo uninstall failed with returncode {0}]".format(res))
+                sys.exit(1)
+            install_cmd = "{0} -m pip install --user {1}".format(py_interp, clingo_pkg)
```
Comment (cyrush) [inline uberenv.py:1082]: since this will be a user install, this breaks the mold of uberenv a bit. But not sure there is an easy alternative.  If the system install already has clingo this path is shortcuted so thats a good option for folks that want to avoid user install.

**Review by cyrush — APPROVED** [review]

File: uberenv.py (hunk 2 above)
Comment (adrienbernede) [reply uberenv.py:787]: I believe you could do
`spack config --scope defaults add config:concretizer:clingo`
and save yourself some trouble here.

**Review by adrienbernede — CHANGES_REQUESTED** [review]
Interesting, I haven't used the new concretizer yet. Testing this will force me to ;)

File: uberenv.py (hunk 2 above)
Comment (cyrush) [reply uberenv.py:787]: using spack config command would avoid the pyyaml req for uberenv, so that's a really good adjustment

**Comment by joshessman-llnl** [comment]
> 
> 
> Interesting, I haven't used the new concretizer yet. Testing this will force me to ;)

The biggest change is that you're required to provide a version for system ("external") packages - there's some discussion at https://github.com/LLNL/serac/pull/417#discussion_r616035913

**Review by cyrush — APPROVED** [review]

**Review by white238 — APPROVED** [review]

File: docs/sphinx/index.rst [hunk 4]
```diff
@@ -181,24 +181,25 @@ is found in the in the following order:
...
+  spack_commit              **None**                   Spack commit to checkout                         **None**
+  spack_activate            **None**                   Spack packages to activate                       **None**
+  spack_configs_path        **None**                   Directory with Spack configs to be copied        ``spack_configs``
+  spack_packages_path       **None**                   Directory with Spack packages to be copied       ``packages``
+  vcpkg_url                 **None**                   Download url for Vcpkg                           ``https://github.com/microsoft/vcpkg``
+  vcpkg_branch              **None**                   Vcpkg branch to checkout                         ``master``
+  vcpkg_commit              **None**                   Vcpkg commit to checkout                         **None**
+  vcpkg_ports_path          ``--vcpkg-ports-path``     Folder with vcpkg ports files                    **None**
+  spack_use_new_concretizer **None**                   Use Spack's ASP solver for concretization        ``False``
```
Comment (kennyweiss) [inline docs/sphinx/index.rst:201]: Future-proofing:  Would it be better to have:
```
spack_concretizer  { 'default', 'clingo', ... }
```

File: uberenv.py (hunk 3 above)
Comment (kennyweiss) [reply uberenv.py:1080]: I tend to agree with @cyrush here.  

Directly calling `pip` on people's python seems different than the typical `uberenv` workflow. It might be preferable to error out here, with a message that tells people how to install `clingo` (and possibly provide some additional documentation about how to bootstrap their environment with `clingo`)

**Comment by joshessman-llnl** [comment]
@adrienbernede's suggestion happens to not work - reported as https://github.com/spack/spack/issues/23259

**Review by kennyweiss — APPROVED** [review]
Thanks @joshessman-llnl 

Approving, but, as noted below, I have concerns about the workaround for getting `clingo`.  It seems like there might be a different/better way to use spack to bootstrap an environment with `clingo` ?

**Comment by cyrush** [comment]
thats a bummer, hopefully they have a quick fix.

**Comment by adrienbernede** [comment]
> @adrienbernede's suggestion happens to not work - reported as [spack/spack#23259](https://github.com/spack/spack/issues/23259)

What version of Spack are you testing with? Posted the command after testing it. I works on develop.
Maybe we have an issue with it in Uberenv context.

**Comment by joshessman-llnl** [comment]
> 
> 
> > @adrienbernede's suggestion happens to not work - reported as [spack/spack#23259](https://github.com/spack/spack/issues/23259)
> 
> What version of Spack are you testing with? I posted the command after testing it. I works on develop.
> Maybe we have an issue with it in Uberenv context.

It looks like the patching Uberenv does with `config.py` is related, as it does seem to work without the patching

File: uberenv.py (hunk 3 above)
Comment (kennyweiss) [reply uberenv.py:1061]: @joshessman-llnl -- I was thinking about this a bit more and wanted to clarify my position/review comments.

This is a tricky problem (i.e. requiring a dependency before you can run the dependency-getting tool). 

While I don't like the tool pip-installing things into the user's python, I'm ok with the proposed solution as a stop-gap until we can find something better. Especially since (a) this is an opt-in solution -- i.e. the user has to specifically ask for `clingo`, and (b) there are issues getting newer versions of spack to concretize without `clingo` in some workflows.


In terms of something better, would it be possible for uberenv to install `clingo` into a `virtualenv`, which it could then use as its python for spack?  My main concern is that installing into a user's system python can have unexpected/undesired side effects.

**tl;dr;** -- I'm ok with the current solution, and will welcome a better solution in the future

**Review by adrienbernede — APPROVED** [review]
Good to go!
